server.js

```diff
@@ -0,0 +1,122 @@
+"use strict";
+
+const finalhandler = require('finalhandler');
+const bodyParser   = require('body-parser');
+const http         = require('http');
+const urlParser    = require('url');
+const querystring  = require('querystring');
+const Router       = require('router');
+const bcrypt       = require('bcrypt');
+
+const router = new Router({ mergerParams: true });
+
+let messages = [];
+let nextId = 1;
+
+class Message {
+  constructor(message) {
+    this.id = nextId;
+    this.message = message;
+    nextId++;
+  }
+}
+
+// Use body parser to handle post data.
+router.use(bodyParser.json());
+
+router.get('/', (request, response) => {
+  response.setHeader('Content-Type', 'text/plain; charset=utf-8');
+  response.end('Hello, World!');
+});
+
+router.get('/messages', (request, response) => {
+  let url = urlParser.parse(request.url),
+      params = querystring.parse(url.query);
+
+  let result = JSON.stringify(messages);
+
+  response.setHeader('Content-Type', 'application/json; charset=utf-8');
+
+  if (params.encrypt) {
+    response.setHeader('Content-Type', 'text/plain; charset=utf-8');
+    return bcrypt.hash(result, 10, (error, hashed) => {
+      if (error) {
+        throw new Error();
+      }
+      response.end(hashed);
+    });
+  }
+
+  response.end(result);
+});
+
+router.get('/message/:id', (request, response) => {
+  let url    = urlParser.parse(request.url),
+      params = querystring.parse(url.query),
+      result, found;
```

Comment:
This is an interesting style choice. I like declaring variables at the top of their scope, but I wonder if it might be clearer if `result` and `found` are declared with `const` where they're first used below?